pkg/networkaware/topologicalsort/topologicalsort.go

```diff
@@ -0,0 +1,136 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package topologicalsort
+
+import (
+	"fmt"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/scheduler/framework"
+
+	agv1alpha "github.com/diktyo-io/appgroup-api/pkg/apis/appgroup/v1alpha1"
+	aglisters "github.com/diktyo-io/appgroup-api/pkg/generated/listers/appgroup/v1alpha1"
+
+	pluginconfig "sigs.k8s.io/scheduler-plugins/apis/config"
+	networkawareutil "sigs.k8s.io/scheduler-plugins/pkg/networkaware/util"
+	"sigs.k8s.io/scheduler-plugins/pkg/qos"
+)
+
+const (
+	// Name : name of plugin used in the plugin registry and configurations.
+	Name = "TopologicalSort"
+)
+
+// TopologicalSort : Sort pods based on their AppGroup and corresponding microservice dependencies
+type TopologicalSort struct {
+	handle     framework.Handle
+	agLister   aglisters.AppGroupLister
+	namespaces []string
+}
+
+// Name : returns the name of the plugin.
+func (ts *TopologicalSort) Name() string {
+	return Name
+}
+
+// getArgs : returns the arguments for the TopologicalSort plugin.
+func getArgs(obj runtime.Object) (*pluginconfig.TopologicalSortArgs, error) {
+	TopologicalSortArgs, ok := obj.(*pluginconfig.TopologicalSortArgs)
+	if !ok {
+		return nil, fmt.Errorf("want args to be of type TopologicalSortArgs, got %T", obj)
+	}
+
+	return TopologicalSortArgs, nil
+}
+
+// New : create an instance of a TopologicalSort plugin
+func New(obj runtime.Object, handle framework.Handle) (framework.Plugin, error) {
+	klog.V(4).InfoS("Creating new instance of the TopologicalSort plugin")
+
+	args, err := getArgs(obj)
+	if err != nil {
+		return nil, err
+	}
+
+	agLister, err := networkawareutil.InitAppGroupInformer(handle.KubeConfig())
+	if err != nil {
+		return nil, err
+	}
+
+	pl := &TopologicalSort{
+		handle:     handle,
+		agLister:   agLister,
+		namespaces: args.Namespaces,
+	}
+	return pl, nil
+}
+
+// Less is the function used by the activeQ heap algorithm to sort pods.
+// 1) Sort Pods based on their AppGroup and corresponding service topology graph.
+// 2) Otherwise, follow the strategy of the QueueSort Plugin
+func (ts *TopologicalSort) Less(pInfo1, pInfo2 *framework.QueuedPodInfo) bool {
+	p1AppGroup := networkawareutil.GetPodAppGroupLabel(pInfo1.Pod)
+	p2AppGroup := networkawareutil.GetPodAppGroupLabel(pInfo2.Pod)
+
+	// If pods do not belong to an AppGroup, or being to different AppGroups, follow vanilla QoS Sort
+	if p1AppGroup != p2AppGroup || len(p1AppGroup) == 0 {
+		klog.V(4).Infof("Pods do not belong to the same appGroup: %v and %v, following Qos strategy...", p1AppGroup, p2AppGroup)
+		s := &qos.Sort{}
+		return s.Less(pInfo1, pInfo2)
+	}
+
+	klog.V(4).InfoS("Pods: %v and %v from appGroup %v", pInfo1.Pod.Name, pInfo2.Pod.Name, p1AppGroup)
```

Comment:
This functin is on the hot path, and then hence this can be a frequent call, so maybe remove this log or increase the verbose level to V(5).